### DISCUSSION TITLE
Unable to generate a response action using catchToEffect().map(...)

### DISCUSSION THREAD

**Discussion by kaduva** [pointfreeco/swift-composable-architecture#468] [Q&A] [answered]
Hi, thanks for this amazing library.

I am trying to login to application which uses Firebase and after a successful login I want another action to be generated like below

```
let authenticationReducer = Reducer<AuthenticationState, AuthenticationAction, AuthenticationEnvironment> {
    state, action, environment in
    switch(action) {
    case .authenticationResponse(.failure):
        state.user = nil
        return .none
    case let .authenticationResponse(.success(response)):
        state.user = response
        return .none
    case let .authenticate(email, password):
        struct AuthenticationId: Hashable {}
        state.email = email
        environment
            .authenticationClient
            .authenticateWithEmail(email, password)
            .receive(on: environment.mainQueue)
            .catchToEffect()
            .map(AuthenticationAction.authenticationResponse) // <- this seems to be NOT working
            .cancellable(id: AuthenticationId(), cancelInFlight: true) // <- Result of call to 'cancellable(id:cancelInFlight:)' is unused - not sure why
        return .none
    }
}
```
I can see that authenticate action is passed to reducer as expected and promise results in a .success(user) callback. But  / `.catchToEffect().map(AuthenticationAction.authenticationResponse)` not working. 

The method definition is as follows:

```

extension AuthenticationClient {
    static let live = AuthenticationClient { (email, password) -> Effect<User, AuthenticationClient.Failure> in
        Future<User, AuthenticationClient.Failure>{ promise in
            Auth.auth().createUser(withEmail: email, password: password) { authResult, error in
                if let authResult = authResult {
                    let user = User(uuid: authResult.user.uid, email: authResult.user.email ?? "")
                    print(user)
                    return promise(.success(user))
                }
                return promise(.failure(AuthenticationClient.Failure()))
            }
        }.eraseToEffect();
    }
}
```

And action

```
enum AuthenticationAction: Equatable {
    case authenticationResponse(Result<User, AuthenticationClient.Failure>)
    case authenticate(String, String)
}
```

I am new swift and iOS platform - any help will be greatly appreciated.

**Comment by fabstu** [comment, marked as answer]
> return .none

You might want to return the created effect instead. Does that fix it?

About cancellable: I'm new to this too, but looking at its source code I think it works differently thank you might think. I think it creates a new effect and redirects all messages from the already existing one to the new one. On .cancel the newly created effect gets canceled, so you essentially discard the results of the authentication call, but that does not stop that authentication call, whatever the result may be.

You might have already been aware of this.

**Reply by kaduva** [reply]
Yes, returning the call fixes it, thank you.

**Reply by Dominic263** [reply]
Hi Kaduva, would you be open to sharing how you structured your full Authentication client?  I am also running into issues with Firebase and would love to see how other folks implement authentication. Thanks

**Reply by mbrandonw** [reply]
@Dominic263 this would also be great to split out into a new discussion, because I'm sure others have a similar question and/or experience with Firebase!
